decArgo_api/decoder_bindings/zip_nc_files.py

```diff
@@ -0,0 +1,58 @@
+"""Code to Zip decoded NC files."""
+
+import logging
+import traceback
+import uuid
+import xml.etree.ElementTree as ET
+from contextlib import contextmanager
+from io import BytesIO
+from pathlib import Path
+from zipfile import ZIP_STORED, ZipFile
+
+
+class ZipNCFilesError(Exception):
+    """Raised when there is an issue zipping the NC files."""
+
+
+class ZipNCFiles:
+    """Methods to zip newly decoded NC files from the Coriolis Decoder."""
+
+    def __init__(self, wmonum: str, xml_location: str = "/mnt/data/output/xml"):
+        """Initialise the instance with the wmonum."""
+        self.wmonum = wmonum
+        self.xml_location = xml_location
+
+    @contextmanager
+    def bytes_buffer(self):
+        """Provide an open BytesIO object for the zipfile creation."""
+        with BytesIO() as bytes_io:
+            yield bytes_io
+
+    def fetch_all_nc_paths(self) -> list[str]:
+        """Parse the XML and return a list of paths generated by the decoder."""
+        latest_xml_file = max(Path(self.xml_location).iterdir(), key=lambda file: file.stat().st_mtime)
+        with open(latest_xml_file) as xml_file:
+            root = ET.fromstring(xml_file.read())
+            return [item.text for item in root.iter() if item.tag.startswith("output")]
+
+    def zip_all_nc_files(self) -> BytesIO:
+        """Zip all decoded NC files and return to the app so FastAPI can stream it."""
+        try:
+            with self.bytes_buffer() as temp_bytes:
+                with ZipFile(temp_bytes, "w", compression=ZIP_STORED) as zipfile:
+                    for file in self.fetch_all_nc_paths():
+                        zipfile.write(file)
+
+                temp_bytes.seek(0)
+                data = temp_bytes.read()
+
+            buffer = BytesIO(data)
+            buffer.seek(0)
+            zip_filename = f"{self.wmonum}_{uuid.uuid4()}.zip"
+            return buffer, zip_filename
```

Comment:
I'm thinking that it would be better to turn this method into a context manager and return just the buffer – you can generate the filename in whatever function is calling this.